### PR TITLE
Adding CH376 chip emulation for Orix/telemon30

### PR DESCRIPTION
**Pull request by jedeoric**
Hello, as explained on defence-force forum, here is the ch376 emulation.

This emulation is for telestrat only, because i did not finished the electronic card for the atmos, but telestrat mode on Oricutron will do the job.

I did some "ifdef WIN32" because i did emulation on windows. For linux, it should work, but someone should test without ifdef. Morphos should work too because emulation works on Morphos, but i did not compile oricutron on Morphos.

Other thing, when you start telestrat, ch376 emulation starts. If it's a problem let me know.

I would like to add telemon3.0 roms and Orix roms, but i don't know where the rom are in this repository. How can i share theses ROMs in oricutron ?

I did a plugins folder because i think that it's better to prepare any plugins management. And it should be clear in this way. Anyway, if it's a problem let me know.

Docs are here : http://orix.oric.org

Usbdrive is the main folder. Emulation will read files in this directory.

If something is wrong let me know :)

I would like to add some binary on usbdrive folder, let me know if it's a problem for you

### CODE REVIEW

File: machine.c [hunk 1]
```diff
@@ -610,6 +626,11 @@ unsigned char telestratread( struct m6502 *cpu, unsigned short addr )
   {
     switch( addr & 0x0f0 )
     {
+#ifdef WIN32
+      case 0x040:
+        return ch376_oric_read(oric->tele_ch376, addr);
+        break;
```
Comment (polluks) [inline machine.c]: dead code because of return

**Comment by jedeoric** [comment]
Corrected real ch376 page 3 address in machine.c as Polluks noticed

File: plugins/ch376/oric_ch376_plugin.h [hunk 2]
```diff
@@ -0,0 +1,8 @@
+struct ch376 * ch376_oric_init(void);
+void ch376_oric_reset(struct ch376 *ch376);
+void	ch376_oric_write(struct  ch376 *ch376, Uint16 addr, Uint8 data);
+unsigned char 	ch376_oric_read(struct  ch376 *ch376, Uint16 addr);
+void ch376_oric_config(struct ch376 *ch376);
+void ch376_oric_destroy(struct ch376 *ch376);
+
+void ch376_oric_reset(struct ch376 *ch376);
```
Comment (polluks) [inline plugins/ch376/oric_ch376_plugin.h]: EOL

**Comment by jedeoric** [comment]
Corrected :)

File: plugins/ch376/ch376.c [hunk 3]
```diff
@@ -0,0 +1,2039 @@
...
+}
+
+static void system_free_mem(void *ptr)
+{
+    free(ptr);
+}
+
+static CH376_BOOL system_init_context(CH376_CONTEXT *context, UNUSED void *user_data)
+{
+  /* Nothing to do */
```
Comment (assinie) [inline plugins/ch376/ch376.c]: Need: return CHS376_TRUE;
or ch376_create() will return NULL when compiled on Linux

File: plugins/ch376/ch376.c [hunk 4]
```diff
@@ -0,0 +1,2039 @@
...
+}
+
+static CH376_FILE system_file_open_new(CH376_CONTEXT *context, const char *file_name, CH376_LOCK root_lock)
+{
+    return file_open(file_name, root_lock, "wb");
+}
+
+static void system_file_close(CH376_CONTEXT *context, CH376_FILE file)
+{
+    fclose(file);
```
Comment (assinie) [inline plugins/ch376/ch376.c]: must be enclose by:
if (file) { }

File: plugins/ch376/ch376.c [hunk 5]
```diff
@@ -0,0 +1,2039 @@
...
+            is_done = CH376_TRUE;
+        }
+    }
+
+    return is_done;
+}
+
+static void system_finish_examine_directory(CH376_CONTEXT *context, CH376_DIR fib)
+{
+    closedir(fib->handle);
```
Comment (assinie) [inline plugins/ch376/ch376.c]: must be enclosed by:
if (fib) { ... }

File: plugins/ch376/ch376.c (hunk 3 above)
Comment (assinie) [inline plugins/ch376/ch376.c]: Need: return CHS376_TRUE;
or ch376_create() will return NULL when compiled on Linux

File: plugins/ch376/ch376.c (hunk 4 above)
Comment (assinie) [inline plugins/ch376/ch376.c]: must be enclose by:
if (file) { }

File: plugins/ch376/ch376.c (hunk 5 above)
Comment (assinie) [inline plugins/ch376/ch376.c]: must be enclosed by:
if (fib) { ... }

File: machine.c [hunk 6]
```diff
@@ -1091,6 +1114,10 @@ SDL_bool emu_event( SDL_Event *ev, struct machine *oric, SDL_bool *needrender )
           via_init( &oric->via, oric, VIA_MAIN );
           via_init( &oric->tele_via, oric, VIA_TELESTRAT );
           acia_init( &oric->tele_acia, oric );
+#ifdef WIN32
+          oric->tele_ch376=ch376_oric_init();
+          ch376_oric_config(oric->tele_ch376);
```
Comment (assinie) [inline machine.c]: Maybe add a test to check the value of oric->tele_ch376 against NULL in case of memory allocation or context initialization failure in function ch376_create() called by ch376_oric_init()

**Comment by pete-gordon** [comment]
> I would like to add telemon3.0 roms and Orix roms, but i don't know where the rom are in this repository. How can i share theses ROMs in oricutron ?

Original Oric ROMs are not present in the git repository, because ownership and licensing are unclear. I take the risk of including them in the binary releases, but it is not appropriate for them to be in the repo.

If the Orix ROM is yours, you are welcome to add it to the repo.

**Comment by pete-gordon** [comment]
Please change the "#ifdef WIN32" to a check for "oric->ch376enabled" (or so), and add it as an option to the hardware options menu (and commandline).

File: plugins/ch376/ch376.c (hunk 3 above)
Comment (jedeoric) [inline plugins/ch376/ch376.c]: fixed

File: plugins/ch376/ch376.c (hunk 4 above)
Comment (jedeoric) [inline plugins/ch376/ch376.c]: fixed

File: plugins/ch376/ch376.c (hunk 5 above)
Comment (jedeoric) [inline plugins/ch376/ch376.c]: fixed

File: machine.c (hunk 6 above)
Comment (jedeoric) [inline machine.c]: fixed

**Comment by jedeoric** [comment]
@pete-gordon  I added Orix rom managed in gitignore. 
For "oric->ch376enabled", i prefer to manage it by configuration file, are you ok ?

**Comment by pete-gordon** [comment]
Yes, that's fine. We can add menu items and command line options later.

**Comment by jedeoric** [comment]
He will try to add menu items.

**Comment by jedeoric** [comment]
Hello,

I added ch376 into the menu in hardware section below serial option. ch376 option is also added in .cfg
it's working.
Please note that i did the code to have ch376 working on telestrat emulation. ch376 will work easily on atmos but i don't like the idea to activate the chip in atmos, when i did not released my new card with others stuff on the Atmos

**Comment by pete-gordon** [comment]
What you can do is make it so that if you select "ch376" it automatically switches to telestrat mode. That's what happens when you select other impossible configs, it just switches to the right combination. Do you want to do that on this pull request, or shall I take this one now?

File: machine.c [hunk 7]
```diff
@@ -342,7 +350,17 @@ void telestratwrite( struct m6502 *cpu, unsigned short addr, unsigned char data
...
     {
-      case 0x20:
+
+      case 0x40:
+        if (oric->ch376_activated)
+        {
+          if (addr == 0x340 || addr == 0x341)
+            ch376_oric_write(oric->ch376, addr, data);
+        }
+      break;
```
Comment (pete-gordon) [inline machine.c]: This isn't quite correct. If ch476 is NOT activated, writes to 0x34x do nothing, instead of the normal VIA write.

Please move the 0x40 case to just before the "default:" and only do "break" inside the if(), that way if ch376 is NOT enabled, it'll drop into the default case.

File: machine.c [hunk 8]
```diff
@@ -610,6 +628,14 @@ unsigned char telestratread( struct m6502 *cpu, unsigned short addr )
...
     switch( addr & 0x0f0 )
     {
+      case 0x040:
+        if (oric->ch376_activated)
+        {
+          if (addr == 0x340 || addr == 0x341)
+            return ch376_oric_read(oric->ch376, addr);
+        }
+        break;
+
```
Comment (pete-gordon) [inline machine.c]: Same as the comment above.

**Comment by jedeoric** [comment]
hello, for telestrat switch, i prefer to release my card before, and i'll add ch376 for others machines.

For the bug noticed, i totally forgot this (VIA6522 answer on the whole page 3 when there is no hardware behind.

**Comment by pete-gordon** [comment]
Thanks :-)
